ExtraMojo/bstr/bstr.mojo

```diff
@@ -0,0 +1,198 @@
+import math
+from algorithm import vectorize
+from collections import Optional
+from memory import Span, UnsafePointer
+from sys.info import simdwidthof
+
+
+# TODO: split this all out and create similar abstractions as the Rust bstr crate
+# TODO: add an ascii to lower case
+
+
+alias SIMD_U8_WIDTH: Int = simdwidthof[DType.uint8]()
+
+
+@always_inline
+fn arg_true[simd_width: Int](v: SIMD[DType.bool, simd_width]) -> Int:
+    for i in range(simd_width):
+        if v[i]:
+            return i
+    return -1
+
+
+@always_inline
+fn find_chr_all_occurrences(haystack: Span[UInt8], chr: UInt8) -> List[Int]:
+    """Find all the occurrences of `chr` in the input buffer."""
+    var holder = List[Int]()
+
+    if len(haystack) < SIMD_U8_WIDTH:
+        for i in range(0, len(haystack)):
+            if haystack[i] == chr:
+                holder.append(i)
+        return holder
+
+    @parameter
+    fn inner[simd_width: Int](offset: Int):
+        var simd_vec = haystack.unsafe_ptr().load[width=simd_width](offset)
+        var bool_vec = simd_vec == chr
+        if bool_vec.reduce_or():
+            # TODO: @unroll
+            for i in range(len(bool_vec)):
+                if bool_vec[i]:
+                    holder.append(offset + i)
+
+    vectorize[inner, SIMD_U8_WIDTH](len(haystack))
+    return holder
+
+
+@always_inline
+fn find_chr_next_occurrence_simd(
+    haystack: Span[UInt8], chr: UInt8, start: Int = 0
+) -> Int:
+    """
+    Function to find the next occurrence of character using SIMD instruction.
+    The function assumes that the tensor is always in-bounds. any bound checks should be in the calling function.
+    """
+    if len(haystack) < SIMD_U8_WIDTH:
+        for i in range(0, len(haystack)):
+            if haystack[i] == chr:
+                return i
+        return -1
+
+    var haystack_len = len(haystack) - start
+    var aligned = start + math.align_down(haystack_len, SIMD_U8_WIDTH)
+
+    for s in range(start, aligned, SIMD_U8_WIDTH):
+        var v = haystack[s:].unsafe_ptr().load[width=SIMD_U8_WIDTH]()
+        var mask = v == chr
+        if any(mask):
+            return s + arg_true(mask)
+
+    for i in range(aligned, len(haystack)):
+        if haystack[i] == chr:
+            return i
+
+    return -1
+
+
+alias CAPITAL_A = SIMD[DType.uint8, SIMD_U8_WIDTH](ord("A"))
+alias CAPITAL_Z = SIMD[DType.uint8, SIMD_U8_WIDTH](ord("Z"))
+alias TO_LOWER_DIFF = SIMD[DType.uint8, SIMD_U8_WIDTH](ord("a") - ord("A"))
+alias ZERO = SIMD[DType.uint8, SIMD_U8_WIDTH](0)
+
+
+@always_inline
+fn to_ascii_lowercase_simd(mut buffer: List[UInt8]):
```

Comment:
This is not nearly as fast as I'd hoped. See https://lemire.me/blog/2024/08/03/converting-ascii-strings-to-lower-case-at-crazy-speeds-with-avx-512/